ftdc/ftdc.go

```diff
@@ -151,14 +153,47 @@ func (ftdc *FTDC) Remove(name string) {
 
 	for idx, statser := range ftdc.statsers {
 		if statser.name == name {
-			ftdc.logger.Debugw("Removed statser", "name", name, "type", fmt.Sprintf("%T", statser.statser))
+			ftdc.logger.Debugw("Removed statser", "name", name,
+				"type", fmt.Sprintf("%T", statser.statser), "generationId", ftdc.inputGenerationID)
 			ftdc.statsers = slices.Delete(ftdc.statsers, idx, idx+1)
 			ftdc.inputGenerationID++
 			return
 		}
 	}
 
-	ftdc.logger.Warnw("Did not find statser to remove", "name", name)
+	ftdc.logger.Warnw("Did not find statser to remove",
+		"name", name, "generationId", ftdc.inputGenerationID)
+}
+
+// conditionalRemoveStatser first checks the generation matches before removing the `name` Statser.
```

Comment:
I frankly didn't need to use this technique here. But I felt I'd end up then justifying what the user contract would have to be to avoid a "bug". And given the expected usage is pretty basic, I did this instead?

Just take this as an opportunity to see optimistic concurrency control in action. Understand why we need something in addition to the locks. Ask yourself how this might change if we instead did validation on `constructDatum`. And hypothesize why that might be undesirable.

Or don't. It's a good thing for you two to know about (I know this isn't exactly the first time you're reasoning about it -- but I suspect it's not immediately obvious what could go "wrong" without this). But this doesn't have to be the when and why for that.